myokit/formats/sbml/_importer.py

```diff
@@ -35,334 +36,1065 @@ def __init__(self):
+                ' No model provided.')
+
+        # Get model name
+        name = self._getName(SBMLmodel)
+        if not name:
             name = 'Imported SBML model'
 
         # Create myokit model
         model = myokit.Model(self._convert_name(name))
         log.log('Reading model "' + model.meta['name'] + '"')
 
-        # Create one giant component to hold all variables
-        comp = model.add_component('sbml')
-
-        # Create table of variable names
-        refs = {}
-
-        # Handle notes, if given
-        x = xmodel.find(self.ns['sbml'] + 'notes')
-        if x:
+        # Add notes, if provided, to model description
+        notes = self._getNotes(SBMLmodel)
+        if notes:
             log.log('Converting <model> notes to ascii')
-            model.meta['desc'] = html2ascii(
-                ET.tostring(x).decode(), width=75)
+            model.meta['desc'] = html2ascii(notes, width=75)
             # width = 79 - 4 for tab!
 
-        # Warn about missing functionality
-        x = xmodel.find(self.ns['sbml'] + 'listOfCompartments')
-        if x:   # pragma: no cover
-            log.warn('Compartments are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfSpecies')
-        if x:   # pragma: no cover
-            log.warn('Species are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfConstraints')
-        if x:   # pragma: no cover
-            log.warn('Constraints are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfReactions')
-        if x:   # pragma: no cover
-            log.warn('Reactions are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfEvents')
-        if x:   # pragma: no cover
-            log.warn('Events are not supported.')
-
-        # Ignore custom functions
-        x = xmodel.find(self.ns['sbml'] + 'listOfFunctionDefinitions')
-        if x:   # pragma: no cover
-            log.warn('Custom math functions are not (yet) implemented.')
-
-        # Parse custom units
-        x = xmodel.find(self.ns['sbml'] + 'listOfUnitDefinitions')
-        if x:
-            self._parse_units(model, comp, x)
-
-        # Add time as independent variable (not explicit in SBML format)
-        if bind_time:
-            # Add and bind time variable to component
-            time = comp.add_variable_allow_renaming('time')
-            time.set_binding('time')
-
-            # Set unit and value
+        # Raise error if function definitions are provided (could be added in
+        # another PR)
+        funcDefs = self._getListOfFunctionDefinitions(SBMLmodel)
+        if funcDefs:
+            raise SBMLError(
+                'Myokit does not support functionDefinitions. Please insert '
+                'your function wherever it occurs in yout SBML file and delete'
+                ' the functionDefiniton in the file.')
+
+        # Create user defined unit reference
+        self.userUnitDict = dict()
```

Comment:
Is this meant to be accessed by the user? (And if so, is that safe?) Otherwise make private

Reply:
Also, it seems you're building up some state in this parsing method, and storing it in the Parser instance. That's fine, but to make this re-usable I would suggest to either:

1. Split the parsing into several functions, that are each called with all the info they need. So any function that needs the userUnitDict would have it as an argument. This is my preferred option.

2. Alternatively, wrap all calls to `_parse_model` in a try/finally block that resets the instance back to its original state. So in `def parse()` you would have
```
try:
    self._parse_model(..)
...
finally:
    del self.userUnitDict
    del self.whatever
```

I'd try and do 1, and then where you really really can't avoid it use 2 as well